### ISSUE TITLE
No Idea How To Run 0.5.0 With WSL

### ISSUE THREAD

**Issue by skjbulcher** [neovide/neovide#693]
The documentation says 

> Neovide supports displaying a full gui window from inside wsl via the --wsl command argument.

However, I have no idea how to make that work. I downloaded both the `neovide` and `neovide.exe` 0.5.0 files from https://github.com/Kethku/neovide/releases. 

- The .exe ran fine in Windows, but doesn't start if I provide a `--wsl` argument via a shortcut or from the command line.
- Running `./neovide --wsl` from WSL (Ubuntu) responds with "nvim: Unknown option argument: "--wsl"`.
- Running `./neovide.exe`  from WSL launches neovide in windows, but it isn't much different from running it from windows directly. All paths are Windows paths, and there is no sense that it is using the "wsl copy of neovim" as the documentation claims.

The instructions leave me clueless. I don't use visualstudio, and I'm not about to install and set up VS to work with WSL just to get an idea about how neovide _might_ work.

I'd like to explore neovide to see if it could work for me, but at this point I don't see how I can?

neovide: v0.5.0
neovim: v0.4.3
This issue has to do with running neovide, not neovim.

EDIT: Clarify that this issue is about neovide and not neovim, since both have a v0.5.0.

**Comment by Kethku** [comment]
I'm happy to help you out. The commands have just changed significantly and I'm guessing the docs haven't caught up.

Some tips:
1. You have to have neovim installed in wsl for it to work. You have to run the neovide exe from the windows side and it will try to start up neovim via the nvim.exe on the wsl side.
2. If that doesn't work, you can also start up neovim in wsl via tcp socket. There are some details here: https://github.com/Kethku/neovide/tree/opengl#remote-tcp-support although this is currently on the opengl branch. I'm going to fix the CI build so you can download a current build of that tonight.
3. If all that fails, you could try running with --log to get a log file written out to disk which might give some further hints.

If any of that works let me know, I'd like to update the docs to make it clearer. If not, I've got some other things to try. Please be patient. Its just me and a few other folks building this and supporting mac, linux, and windows is not trivial especially when using a language as new as rust.

**Comment by skjbulcher** [comment]
Good to know I wasn't missing something!

> You have to have neovim installed in wsl for it to work. You have to run the neovide exe from the windows side and it will try to start up neovim via the nvim.exe on the wsl side.

How does neovide.exe know to start up neovim in WSL? Here's what I'm thinking:

I have WSL running, and neovim can be run from the WSL command line (an Ubuntu instance) via "vi", "vim", or "nvim" (there is no "nvim.exe" in WSL to my knowledge). I want to run neovide, say to edit files in my cwd. What do I do? Normally I would simply run `vi [filename]`, or even just `vi`.

What you said suggests that if I run neovide.exe on windows, it will automagically connect to and run from WSL, but how does it know what directory I'm in or what file I want to edit? I still think I'm missing a big piece of the puzzle.

**Comment by skjbulcher** [comment]
I tried the Remote TCP Support documentation without success as well. I ran the nvim headless command as stated in the documentation, and I ran the remote-tcp option as well (both from WSL and from a shortcut I created with the TCP options). Here's an image from my terminal session running byobu, where you can see both commands:

![image](https://user-images.githubusercontent.com/16672299/119904247-e7829f00-befe-11eb-9926-cb1ff9e3848f.png)

neovide did not start up when I provided the --remote-tcp option. It starts fine without it (but of course I'm not connecting to nvim in WSL).

**Comment by skjbulcher** [comment]
Sorry for stringing my replies out; I'm trying your options one at a time.

using the --log option gives me this with the --remote-tcp command:

```
INFO [neovide::bridge] Close watcher started
ERROR [neovide::bridge] Neovide requires version 0.4 or higher
```

NVIM is at v.0.4.3, and neovide is at version 0.5.0.


I wonder if part of my confusion is that I don't know what the different released are for? I've been assuming that "neovide" is the linux binary, "Neovide.app.zip" is an iOS binary, and "neovide.exe" is the Windows binary. I don't know if I only need the Windows binary, or if I need something else.

**Comment by skjbulcher** [comment]
On a hunch, I uninstalled a windows version of neovim I had installed. Now when I run `neovide.exe --log`, I see the following:

```
ERROR [neovide::error_handling] Could not locate or start the neovim process: The system cannot find the file specified. (os error 2)
INFO [neovide::window] window created
INFO [neovide::window] renderer created
INFO [neovide::window] Starting window event loop
TRACE [neovide::editor::grid] Editor resized
TRACE [neovide::editor::grid] Editor cleared
TRACE [neovide::editor::grid] Editor cleared
TRACE [neovide::redraw_scheduler] Next frame queued
TRACE [neovide::redraw_scheduler] Next frame queued
TRACE [neovide::redraw_scheduler] Next frame queued
...
DEBUG [neovide::window] Render Triggered
DEBUG [neovide::window] Render Triggered
TRACE [neovide::bridge] UiCommand queued: MouseButton { action: "press", position: (57, 20) }
ERROR [neovide::error_handling] Could not send UI command from the window system to the neovim process.: channel closed
```

When I click a button the window closes. It doesn't appear to be able to find neovim.

**Comment by PyGamer0** [comment]
> How does neovide.exe know to start up neovim in WSL? 

try opening neovide like:
```
./neovide.exe --wsl
```

**Comment by skjbulcher** [comment]
> > How does neovide.exe know to start up neovim in WSL?
> 
> try opening neovide like:
> 
> ```
> ./neovide.exe --wsl
> ```

As Kethku [mentioned](https://github.com/Kethku/neovide/issues/693#issuecomment-849962967), that doesn't work on version 0.5.0

**Comment by Kethku** [comment]
I think maybe there are some misunderstandings that I need to clear up in the readme.

You mentioned wanting to use vi, or vim on the wsl side. Neovide only supports neovim though. SO when you run neovide with the wsl argument, it looks up the location of the installed neovim executable in the wsl filesystem. If it doesn't find one, it will just fail to run, but if it does, it will start up the neovim instance and communicate with it over the windows-wsl boundary via standard input output.

I think the issues you are running into have been fixed in the opengl branch. I've been fighting ci build issues all day, and I think I've gotten it hammered out, but I will update this thread when I have a built exe for sure. I recommend trying that rather than keeping on with the main branch/release builds.

**Comment by Kethku** [comment]
Ok a run finally finished: https://github.com/Kethku/neovide/actions/runs/887728507 I recommend downloading the exe from here.

**Comment by skjbulcher** [comment]
> You mentioned wanting to use vi, or vim on the wsl side. Neovide only supports neovim though. SO when you run neovide with the wsl argument, it looks up the location of the installed neovim executable in the wsl filesystem. If it doesn't find one, it will just fail to run, but if it does, it will start up the neovim instance and communicate with it over the windows-wsl boundary via standard input output.

I think I wasn't clear in my previous message when I started talking about using 'vi' and 'vim'. I have those commands aliased to neovim via `update-alternatives`, and I no longer use vim. neovim is installed in WSL, and it is my daily driver. Sorry for the confusion! Here are more details:

```
root@DESKTOP:/$ ll $(which vi)
lrwxrwxrwx 1 root root 20 Aug  4  2020 /usr/bin/vi -> /etc/alternatives/vi*
root@DESKTOP:/$ ll /etc/alternatives/vi
lrwxrwxrwx 1 root root 13 Oct 22  2020 /etc/alternatives/vi -> /usr/bin/nvim*
root@DESKTOP:/$ ll $(which vim)
lrwxrwxrwx 1 root root 21 Aug  4  2020 /usr/bin/vim -> /etc/alternatives/vim*
root@DESKTOP:/$ ll /etc/alternatives/vim
lrwxrwxrwx 1 root root 13 Oct 22  2020 /etc/alternatives/vim -> /usr/bin/nvim*
```

> I think the issues you are running into have been fixed in the opengl branch. I've been fighting ci build issues all day, and I think I've gotten it hammered out, but I will update this thread when I have a built exe for sure. I recommend trying that rather than keeping on with the main branch/release builds.
>
> Ok a run finally finished: https://github.com/Kethku/neovide/actions/runs/887728507 I recommend downloading the exe from here.

I tried running the executable from that link, but it exited right away. When I ran it with `--log`, it created a file with one line in it:

```
ERROR [neovide::bridge] nvim not found!
```

As I'm beginning to understand how neovide works, I think a few more details about my setup might be helpful. Here is the output of `wsl --list`:

```
> wsl --list
Windows Subsystem for Linux Distributions:
Ubuntu-20.04 (Default)
```

I hope that helps! I'm running an older version of neovide one where the `--wsl` option works, and I'm really liking it so far - good job! It's a bit slow to start up though, so I can't use it as a replacement for nvim, only for dedicated editing sessions. I find that the graphics help me confirm when I run some habitual commands, such as `:w`, because the cursor jumping to the console is evident in my peripheral vision.

**Comment by horseinthesky** [comment]
Hello.
```
D:\>neovide.exe --wsl
```
works great.

Is there an option to set the path to nvim executable?
I'm using both stable(`nvim`) and nightly(`nvim-nightly`) version and `neovide` always runs stable version.

**Comment by skjbulcher** [comment]
>Is there an option to set the path to nvim executable?
> I'm using both stable(nvim) and nightly(nvim-nightly) version and neovide always runs stable version.

Please don't hijack this issue @horseinthesky. If you have a question, please start a new issue so that this issue stays on track, and so that others can find help when they have similar issues. It's hard to track an issue in an issue tracker when threads are treated like a forum.

It could be that the issue is more specific than the title suggests, but so far I don't have any reason to think otherwise.

> ```
> D:\>neovide.exe --wsl
> ```
> 
> works great.

It still does not for me (I just re-downloaded v0.5.0 and tested it again). Note that it does work on older versions for me, I do have a few questions if it is working for you though:

- What version of neovide are you running with `--wsl`?
- What version of linux do you run in WSL?

On a side note, I'm not sure that v0.5.0 is supposed to use the `--wsl` argument, based on the [author's feedback](https://github.com/Kethku/neovide/issues/693#issuecomment-849962967):

> The commands have just changed significantly
> ...
>You have to have neovim installed in wsl for it to work. You have to run the neovide exe from the windows side and it will try to start up neovim via the nvim.exe on the wsl side.

**Comment by horseinthesky** [comment]
Didn't want to hijack the issue. I think my problem also relates to the issue since I'm not able to run wsl NVIM 0.5.0 with neovide.

```
D:\>neovide.exe --wsl
```
works with 0.4.3 which I have in `/usr/bin/nvim`.

My 0.5.0 is located in `/usr/local/bin/nvim-nighly`. Even if I set `alias nvim='/usr/local/bin/nvim-nighly'` in my wsl `.zshrc`, neovide still runs 0.4.3.
This is why I am asking if it is possible to set up an nvim binary path for neovide.

**Comment by skjbulcher** [comment]
Oh, OK. I think there's some confusion due to an unfortunate conflation of version numbers. I wasn't even thinking about the next release of neovim.

I'm talking about being unable to run neovide 0.5.0. I have neovim 0.4.3 installed in my WSL instance, and neovide quits right away unless I use an older version of neovide with the `--wsl` option.

I'll try to clarify my original post to avoid future confusion.

**Comment by horseinthesky** [comment]
Then in terms of NVIM version, we are alike - it is 0.4.3.

I've just checked
neovide (0.5.0):
```
neovide --wsl
```
quits right away.

neovide (0.7.0) - Smooth Scrolling Beta:
```
neovide --wsl
```
works just fine.
![image](https://user-images.githubusercontent.com/19948672/121227971-ddfd1d80-c894-11eb-9371-33b2dddfd990.png)

**Comment by horseinthesky** [comment]
> That's interesting to me. What distribution(s) are you using for WSL? Here's mine:
I am on Windows 21H1 (Build 19043.1052) and also Ubuntu-20.04

> I'm not sure if it's worth mentioning or not, but I'm running the neovide.exe in Windows, and I don't have the executable in my path (so I have to run it from the directory the executable is in).
The same for me. Just run it from my D:\ disk with some stuff out of the PATH.
  
> On a side note, I have a guess as to why neovide isn't picking up your neovim v0.5.0 install though @horseinthesky. I doubt that neovide starts a shell login to check for nvim, in which case it wouldn't pick up the alias you set up. Instead. you could try a symlink, such that
> 
> ```
> /usr/bin/nvim -> /usr/local/bin/nvim-nighly
> ```
This works cuz (I assume) neovide just looking for `nvim` in PATH `over the windows-wsl boundary via standard input output`. It is also probably using the same user as the host (Windows) system.
So would be nice to have an option to give neovide the name of executable or customize the path to it.

**Comment by Kethku** [comment]
> So would be nice to have an option to give neovide the name of executable or customize the path to it.

Either that, or we could try to run actual bash including the bashrc in order to get the correct user configured path to neovim. Honestly I don't know how that might work as I don't use linux that often myself. Especially with the recent changes to enable commandline arguments in th eopengl branch, I don't think it would be that hard...

**Comment by skjbulcher** [comment]
>Either that, or we could try to run actual bash including the bashrc in order to get the correct user configured path to neovim.

How are you currently obtaining the WSL neovim executable?

**Comment by 4ong** [comment]
Had the same issue - the trick that helped me - Ubuntu WSL should be default one - check using 'wsl -l'

**Comment by BatmanAoD** [comment]
I'm running into an issue that is similar or identical. I have installed `neovim` both in the Windows host and in the Ubuntu WSL system. The Windows version is `NVIM v0.5.0`; the WSL version is `NVIM 0.4.3`. I suspect that Neovide is finding the Windows `nvim` even when using the `--wsl` argument.

With the above setup, when I run `neovide.exe --wsl --log`, I get the behavior described above, and the log reads:

```
INFO [neovide::bridge] Close watcher started
INFO [neovide::window] window created
ERROR [neovide::bridge] Neovide requires version 0.4 or higher
```

When I remove the Windows `nvim.exe` from the WSL `PATH`, leaving only the WSL "native" `nvim` on the path, Neovide opens successfully with the `--wsl` argument. This indicates to me that as long as `nvim.exe` (the Windows version) is on the path, that's what Neovide finds when using `--wsl`, which seems wrong.

However, even when `nvim.exe` has been removed from the path and Neovide has successfully opened, as soon as I press any key or move the window, it closes, and the log says something like:

```
TRACE [neovide::bridge] UiCommand queued: Keyboard("a")
ERROR [neovide::error_handling] Could not send UI command from the window system to the neovim process.: channel closed
```

I'm guessing this should be considered a separate issue, though.
